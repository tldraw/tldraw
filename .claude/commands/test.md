---
description: Create or update vitest tests for the specified file or current context
argument-hint: [file-path]
allowed-tools: Read, Edit, MultiEdit, Bash(yarn test:*), Bash(yarn typecheck:*)
---

# Test Creation and Management

Create comprehensive vitest tests for tldraw codebase following project conventions.

In order get a better idea of what the different components do and how they interact before writing the JSDoc comments, make sure to read the DOCS.md located at packages/[PACKAGE_NAME]/DOCS.md, and/or CONTEXT.md located at packages/[PACKAGE_NAME]/CONTEXT.md (if it exists). Always read these files before starting your work.

## Instructions

**Arguments:**

- `$1` (optional): File path to test (if not provided, uses current context)

## Test Requirements

### File Naming Conventions

- Unit tests: `src/lib/MyFile.test.ts` (same directory as source)
- Integration tests: `src/test/my-feature.test.ts` (separate test directory)

### Test Structure

- Use vitest framework (`import { describe, it, expect } from 'vitest'`)
- Import from relative paths (e.g., `import { atom } from '../Atom'`)
- Group related tests in `describe` blocks with clear names
- Use descriptive test names that explain the behavior being tested (e.g., `'contain data'`, `'can be updated'`)
- For editor tests requiring tldraw shapes/tools: write tests in tldraw workspace, not editor workspace
- Test all exported functions, classes, and methods
- Include edge cases and error conditions
- Test both sync and async operations where applicable
- Use both `it()` and `test()` functions as appropriate (following existing patterns)

### Test Categories

1. **Unit Tests**: Test individual functions/methods in isolation
2. **Integration Tests**: Test interaction between multiple components
3. **Editor Tests**: Test editor functionality (use tldraw workspace for default shapes)

### Running Tests

- Run specific package tests: `yarn test run` from package directory
- Run all tests: `yarn test` from repo root (avoid unless necessary)
- Type checking: `yarn typecheck`

### Test Analysis Strategy

When working with existing tests:

1. **Read existing test files** to understand current coverage and patterns
2. **Compare with source code** to identify gaps in test coverage
3. **Check for outdated assertions** that may no longer match current behavior
4. **Look for missing edge cases** or error scenarios
5. **Evaluate test structure** against tldraw conventions
6. **Assess test performance** and identify potential optimizations

## Your Task

1. **Analyze the target file** to understand its exports and functionality
2. **Check for existing tests** and determine the appropriate action:
   - If no tests exist: Create new test files following naming conventions
   - If tests exist but are incomplete: Add missing test cases and improve coverage
   - If tests exist but are outdated: Update tests to match current implementation
   - If tests are comprehensive: Review and suggest improvements or optimizations

3. **Handle existing test scenarios:**
   - **Missing coverage**: Identify untested functions/methods and add comprehensive tests
   - **Outdated tests**: Update tests that no longer match the current API or behavior
   - **Incomplete test cases**: Add missing edge cases, error conditions, or scenarios
   - **Poor test structure**: Refactor tests to follow tldraw conventions and best practices
   - **Performance issues**: Optimize slow or inefficient test patterns

4. **Write or update tests covering:**
   - All exported functions/classes/methods
   - Happy path scenarios
   - Edge cases and error conditions
   - Type safety where applicable
   - State changes and side effects
   - Transaction handling and rollbacks (where applicable)

5. **Ensure tests follow tldraw testing patterns and best practices:**
   - Use relative imports for local modules
   - Group logically related tests in describe blocks
   - Test both individual functions and their interactions
   - Include async test patterns when needed
   - Maintain consistent naming and structure with existing tests

6. **Validate and finalize:**
   - Run tests to verify they pass
   - Check test coverage is comprehensive
   - Ensure new/updated tests integrate well with existing test suite
   - Always run `yarn typecheck` from the root of the project to ensure no linter errors. Only run it on the specific test file you're working on.

### Tips

If testing editor functionality, remember to use the tldraw workspace for access to default shapes and tools.

When writing tests, NEVER modify the source code that you are testing.

Use test.fails on tests that are failing if you are >75% confident that the tests themselves are correct and that the underlying issue is with the code you're testing. It is possible that the code that you're testing has a bug in it. Do **not** try to fix the bug. Do not try to write your test in a way that causes the test to pass despite the bug. Notify the user that there is a bug you've found,a nd the user will fix it.

We have a linter that will remove any unused imports. You will need to add imports only when they are used.

You can use `npx tsx` to run the code that you are testing.
