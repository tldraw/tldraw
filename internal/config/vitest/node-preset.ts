import path from 'path'
import { defineConfig } from 'vitest/config'

// Equivalent to internal/config/svgTransform.js for Vitest
const svgTransform = {
	name: 'svg-transform',
	transform(_code: string, id: string) {
		if (id.endsWith('.svg')) {
			return {
				code: 'export default {};',
				map: null,
			}
		}
	},
}

export default defineConfig({
	plugins: [svgTransform],
	test: {
		globals: true,
		environment: 'jsdom',
		setupFiles: [path.resolve(__dirname, './setup.ts')],
		include: ['src/**/*.{test,spec}.{js,ts,jsx,tsx}'],
		exclude: [
			'**/test/__fixtures__/**',
			'**/node_modules/**',
			'**/dist/**',
			'**/.tsbuild/**',
			'**/.tsbuild-dev/**',
			'**/.tsbuild-pub/**',
		],
		coverage: {
			include: ['src/**/*.{ts,tsx}'],
			exclude: ['**/*.test.*', '**/*.spec.*', '**/test/**', '**/__tests__/**'],
		},
	},
	resolve: {
		alias: {
			'~': path.resolve(process.cwd(), 'src'),
		},
	},
})
