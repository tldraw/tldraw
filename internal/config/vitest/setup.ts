import { equals, getObjectSubset, iterableEquality, subsetEquality } from '@jest/expect-utils'
import crypto from 'crypto'
import {
	matcherHint,
	printDiffOrStringify,
	printExpected,
	printReceived,
	stringify,
} from 'jest-matcher-utils'
import { TextDecoder, TextEncoder } from 'util'

if (typeof window !== 'undefined') {
	await import('vitest-canvas-mock')
}

// Polyfill for requestAnimationFrame (equivalent to raf/polyfill)
if (typeof globalThis.requestAnimationFrame === 'undefined') {
	globalThis.requestAnimationFrame = (cb: FrameRequestCallback) => {
		return setTimeout(() => cb(Date.now()), 16) as unknown as number
	}
}

if (typeof globalThis.cancelAnimationFrame === 'undefined') {
	globalThis.cancelAnimationFrame = (id: number) => {
		clearTimeout(id)
	}
}

// Global polyfills
global.TextEncoder = TextEncoder
global.TextDecoder = TextDecoder as any
// @ts-expect-error - cannot delete non-optional property
delete global.crypto
global.crypto = crypto as any

// Crypto polyfill (needed for ai package)
if (typeof globalThis.crypto === 'undefined') {
	// eslint-disable-next-line @typescript-eslint/no-require-imports
	const { Crypto } = require('@peculiar/webcrypto')
	globalThis.crypto = new Crypto()
}

// Image.decode polyfill - this should be handled by individual package setups,
// but including it here as a fallback
if (typeof HTMLImageElement !== 'undefined') {
	if (!HTMLImageElement.prototype.decode) {
		HTMLImageElement.prototype.decode = function () {
			return Promise.resolve()
		}
	}
}

function convertNumbersInObject(obj: any, roundToNearest: number): any {
	if (!obj) return obj
	if (Array.isArray(obj)) {
		return obj.map((x) => convertNumbersInObject(x, roundToNearest))
	}
	if (typeof obj === 'number') {
		// || 0 to avoid -0
		return Math.round(obj / roundToNearest) * roundToNearest || 0
	}
	if (typeof obj !== 'object') {
		return obj
	}

	const r: any = {
		__converted: true,
	}

	for (const k of Object.keys(obj)) {
		r[k] = convertNumbersInObject(obj[k], roundToNearest)
	}

	return r
}

expect.extend({
	toCloselyMatchObject(actual: any, expected: any, roundToNearest = 0.0001) {
		const matcherName = 'toCloselyMatchObject'
		const options = {
			isNot: this.isNot,
			promise: this.promise,
		}

		const EXPECTED_LABEL = 'Expected'
		const RECEIVED_LABEL = 'Received'
		const isExpand = (expand?: boolean): boolean => expand !== false

		const newActualObj = convertNumbersInObject(actual, roundToNearest)

		const newExpectedObj = convertNumbersInObject(expected, roundToNearest)

		const pass = equals(newActualObj, newExpectedObj, [iterableEquality, subsetEquality])

		const message = pass
			? () =>
					// eslint-disable-next-line prefer-template
					matcherHint(matcherName, undefined, undefined, options) +
					'\n\n' +
					`Expected: not ${printExpected(expected)}` +
					(stringify(expected) !== stringify(actual)
						? `\nReceived:     ${printReceived(actual)}`
						: '')
			: () =>
					// eslint-disable-next-line prefer-template
					matcherHint(matcherName, undefined, undefined, options) +
					'\n\n' +
					printDiffOrStringify(
						expected,
						getObjectSubset(actual, expected),
						EXPECTED_LABEL,
						RECEIVED_LABEL,
						isExpand(this.expand)
					)

		return { message, pass }
	},
})
