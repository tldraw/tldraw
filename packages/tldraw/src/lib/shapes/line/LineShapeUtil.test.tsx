import {
	IndexKey,
	TLGeoShape,
	TLLineShape,
	createShapeId,
	mockUniqueId,
	sortByIndex,
	structuredClone,
} from '@tldraw/editor'
import { vi } from 'vitest'
import { TestEditor } from '../../../test/TestEditor'
import { TL } from '../../../test/test-jsx'

let nextId = 0
mockUniqueId(() => 'id' + nextId++)

let editor: TestEditor
const id = createShapeId('line1')

vi.useFakeTimers()

beforeEach(() => {
	editor = new TestEditor()
	editor
		.selectAll()
		.deleteShapes(editor.getSelectedShapeIds())
		.createShapes<TLLineShape>([
			{
				id: id,
				type: 'line',
				x: 150,
				y: 150,
				props: {
					points: {
						a1: { id: 'a1', index: 'a1' as IndexKey, x: 0, y: 0 },
						a2: { id: 'a2', index: 'a2' as IndexKey, x: 100, y: 100 },
					},
				},
			},
		])
})

const getShape = () => editor.getShape<TLLineShape>(id)!
const getHandles = () => editor.getShapeHandles<TLLineShape>(id)!

describe('Translating', () => {
	it('updates the line', () => {
		editor.select(id)
		editor.pointerDown(25, 25, { target: 'shape', shape: getShape() })
		editor.pointerMove(50, 50) // Move shape by 25, 25
		editor.expectShapeToMatch<TLLineShape>({
			id: id,
			x: 175,
			y: 175,
		})
	})

	it('updates the line when rotated', () => {
		editor.select(id)

		const shape = getShape()
		editor.updateShape({ ...shape, rotation: Math.PI / 2 })

		editor.pointerDown(250, 250, { target: 'shape', shape: shape })
		editor.pointerMove(300, 400) // Move shape by 50, 150

		editor.expectShapeToMatch<TLLineShape>({
			id: id,
			x: 200,
			y: 300,
		})
	})
})

describe('Mid-point handles', () => {
	it('create new handle', () => {
		editor.select(id)

		editor.pointerDown(200, 200, {
			target: 'handle',
			shape: getShape(),
			handle: getHandles()[1],
		})
		editor.pointerMove(349, 349).pointerMove(350, 350) // Move handle by 150, 150
		editor.pointerUp()

		editor.expectShapeToMatch({
			id: id,
			props: {
				points: {
					a1: { id: 'a1', index: 'a1', x: 0, y: 0 },
					a1V: { id: 'a1V', index: 'a1V', x: 200, y: 200 },
					a2: { id: 'a2', index: 'a2', x: 100, y: 100 },
				},
			},
		})
	})

	it('allows snapping with mid-point handles', () => {
		editor.createShapesFromJsx([<TL.geo x={200} y={200} w={100} h={100} />])

		editor.select(id)

		editor
			.pointerDown(200, 200, {
				target: 'handle',
				shape: getShape(),
				handle: getHandles()[1],
			})
			.pointerMove(198, 230, undefined, { ctrlKey: true })

		expect(editor.snaps.getIndicators()).toHaveLength(1)
		expect(editor.getShapeHandles(id)).toHaveLength(5) // 3 real + 2
		const points = editor.getShape<TLLineShape>(id)!.props.points
		expect(points).toStrictEqual({
			a1: { id: 'a1', index: 'a1', x: 0, y: 0 },
			a1V: { id: 'a1V', index: 'a1V', x: 50, y: 80 },
			a2: { id: 'a2', index: 'a2', x: 100, y: 100 },
		})
	})

	it('allows snapping with created mid-point handles', () => {
		editor.createShapesFromJsx([<TL.geo x={200} y={200} w={100} h={100} />])

		// 2 actual points, plus 1 mid-points:
		expect(getHandles()).toHaveLength(3)

		// use a mid-point handle to create a new handle
		editor
			.select(id)
			.pointerDown(200, 200, {
				target: 'handle',
				shape: getShape(),
				handle: getHandles().sort(sortByIndex)[1]!,
			})
			.pointerMove(230, 200)
			.pointerMove(240, 200)
			.pointerMove(200, 200)
			.pointerUp()

		// 3 actual points, plus 2 mid-points:
		expect(getHandles()).toHaveLength(5)

		// now, try dragging the newly created handle. it should still snap:
		editor
			.pointerDown(200, 200, {
				target: 'handle',
				shape: getShape(),
				handle: getHandles().sort(sortByIndex)[2],
			})
			.pointerMove(198, 230, undefined, { ctrlKey: true })

		expect(editor.snaps.getIndicators()).toHaveLength(1)
		expect(editor.getShapeHandles(id)).toHaveLength(5) // 3 real + 2
		const points = editor.getShape<TLLineShape>(id)!.props.points
		expect(points).toStrictEqual({
			a1: { id: 'a1', index: 'a1', x: 0, y: 0 },
			a1V: { id: 'a1V', index: 'a1V', x: 50, y: 80 },
			a2: { id: 'a2', index: 'a2', x: 100, y: 100 },
		})
	})
})

describe('Snapping', () => {
	beforeEach(() => {
		editor.updateShape({
			id: id,
			type: 'line',
			props: {
				points: {
					a1: { id: 'a1', index: 'a1', x: 0, y: 0 },
					a2: { id: 'a2', index: 'a2', x: 100, y: 0 },
					a3: { id: 'a3', index: 'a3', x: 100, y: 100 },
					a4: { id: 'a4', index: 'a4', x: 0, y: 100 },
				},
			},
		})
	})

	it('snaps endpoints to itself', () => {
		editor.select(id)

		editor
			.pointerDown(0, 0, { target: 'handle', shape: getShape(), handle: getHandles()[0] })
			.pointerMove(50, 95, undefined, { ctrlKey: true })

		expect(editor.snaps.getIndicators()).toHaveLength(1)
		editor.expectShapeToMatch({
			id: id,
			props: {
				points: {
					a1: { id: 'a1', index: 'a1', x: 50, y: 100 },
					a2: { id: 'a2', index: 'a2', x: 100, y: 0 },
					a3: { id: 'a3', index: 'a3', x: 100, y: 100 },
					a4: { id: 'a4', index: 'a4', x: 0, y: 100 },
				},
			},
		})
	})

	it('snaps endpoints to its vertices', () => {
		editor.select(id)

		editor
			.pointerDown(0, 0, { target: 'handle', shape: getShape(), handle: getHandles()[0] })
			.pointerMove(3, 95, undefined, { ctrlKey: true })

		expect(editor.snaps.getIndicators()).toHaveLength(1)
		editor.expectShapeToMatch({
			id: id,
			props: {
				points: {
					a1: { id: 'a1', index: 'a1', x: 0, y: 100 },
					a2: { id: 'a2', index: 'a2', x: 100, y: 0 },
					a3: { id: 'a3', index: 'a3', x: 100, y: 100 },
					a4: { id: 'a4', index: 'a4', x: 0, y: 100 },
				},
			},
		})
	})

	it("doesn't snap to the segment of the current handle", () => {
		editor.select(id)

		editor
			.pointerDown(0, 0, { target: 'handle', shape: getShape(), handle: getHandles()[0] })
			.pointerMove(5, 2, undefined, { ctrlKey: true })

		expect(editor.snaps.getIndicators()).toHaveLength(0)
		editor.expectShapeToMatch({
			id: id,
			props: {
				points: {
					a1: { id: 'a1', index: 'a1', x: 5, y: 2 },
					a2: { id: 'a2', index: 'a2', x: 100, y: 0 },
					a3: { id: 'a3', index: 'a3', x: 100, y: 100 },
					a4: { id: 'a4', index: 'a4', x: 0, y: 100 },
				},
			},
		})
	})

	it('snaps to vertices on other line shapes', () => {
		editor.createShapesFromJsx([
			<TL.line
				x={150}
				y={150}
				points={{
					a1: { id: 'a1', index: 'a1' as IndexKey, x: 200, y: 0 },
					a2: { id: 'a2', index: 'a2' as IndexKey, x: 300, y: 0 },
				}}
			/>,
		])

		editor.select(id)

		const handle = getHandles()[0]
		editor
			.pointerDown(handle.x, handle.y, { target: 'handle', shape: getShape(), handle })
			.pointerMove(205, 1, undefined, { ctrlKey: true })

		expect(editor.snaps.getIndicators()).toHaveLength(1)
		editor.expectShapeToMatch({
			id: id,
			props: {
				points: {
					a1: { id: 'a1', index: 'a1', x: 200, y: 0 },
					a2: { id: 'a2', index: 'a2', x: 100, y: 0 },
					a3: { id: 'a3', index: 'a3', x: 100, y: 100 },
					a4: { id: 'a4', index: 'a4', x: 0, y: 100 },
				},
			},
		})
	})
})

describe('Misc', () => {
	it('preserves handle positions on spline type change', () => {
		editor.select(id)
		const shape = getShape()
		const prevPoints = structuredClone(shape.props.points)

		editor.updateShapes([
			{
				...shape,
				props: {
					spline: 'cubic',
				},
			},
		])

		editor.expectShapeToMatch<TLLineShape>({
			id,
			props: {
				spline: 'cubic',
				points: prevPoints,
			},
		})
	})

	it('resizes', () => {
		editor.select(id)

		editor
			.pointerDown(150, 0, { target: 'selection', handle: 'bottom' })
			.pointerMove(150, 600) // Resize shape by 0, 600
			.expectToBeIn('select.resizing')

		expect(editor.getShape(id)!).toMatchSnapshot('line shape after resize')
	})

	it('nudges', () => {
		editor.select(id)
		editor.nudgeShapes(editor.getSelectedShapeIds(), { x: 1, y: 0 })

		editor.expectShapeToMatch<TLLineShape>({
			id: id,
			x: 151,
			y: 150,
		})

		editor.nudgeShapes(editor.getSelectedShapeIds(), { x: 0, y: 10 })

		editor.expectShapeToMatch<TLLineShape>({
			id: id,
			x: 151,
			y: 160,
		})
	})

	it('align', () => {
		const boxID = createShapeId('box1')
		editor.createShapes([{ id: boxID, type: 'geo', x: 500, y: 150, props: { w: 100, h: 50 } }])

		const box = editor.getShape<TLGeoShape>(boxID)!
		const line = getShape()

		editor.select(boxID, id)

		expect(editor.getShapePageBounds(box)!.maxX).not.toEqual(editor.getShapePageBounds(line)!.maxX)
		editor.alignShapes(editor.getSelectedShapeIds(), 'right')
		vi.advanceTimersByTime(1000)
		expect(editor.getShapePageBounds(box)!.maxX).toEqual(editor.getShapePageBounds(line)!.maxX)

		expect(editor.getShapePageBounds(box)!.maxY).not.toEqual(editor.getShapePageBounds(line)!.maxY)
		editor.alignShapes(editor.getSelectedShapeIds(), 'bottom')
		vi.advanceTimersByTime(1000)
		expect(editor.getShapePageBounds(box)!.maxY).toEqual(editor.getShapePageBounds(line)!.maxY)
	})

	it('duplicates', () => {
		editor.select(id)

		editor.keyDown('Alt').pointerDown(25, 25, { target: 'shape', shape: getShape() })
		editor.pointerMove(50, 50) // Move shape by 25, 25
		editor.pointerUp().keyUp('Alt')

		expect(Array.from(editor.getCurrentPageShapeIds().values()).length).toEqual(2)
	})

	it('deletes', () => {
		editor.select(id)

		editor.keyDown('Alt').pointerDown(25, 25, { target: 'shape', shape: getShape() })
		editor.pointerMove(50, 50) // Move shape by 25, 25
		editor.pointerUp().keyUp('Alt')

		let ids = Array.from(editor.getCurrentPageShapeIds().values())
		expect(ids.length).toEqual(2)

		const duplicate = ids.filter((i) => i !== id)[0]
		editor.select(duplicate)

		editor.deleteShapes(editor.getSelectedShapeIds())

		ids = Array.from(editor.getCurrentPageShapeIds().values())
		expect(ids.length).toEqual(1)
		expect(ids[0]).toEqual(id)
	})
})
