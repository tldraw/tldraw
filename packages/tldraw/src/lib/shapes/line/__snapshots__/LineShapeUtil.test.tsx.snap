// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Misc > resizes > line shape after resize 1`] = `
{
  "id": "shape:line1",
  "index": "a1",
  "isLocked": false,
  "meta": {},
  "opacity": 1,
  "parentId": "page:page",
  "props": {
    "color": "black",
    "dash": "draw",
    "points": {
      "a1": {
        "id": "a1",
        "index": "a1",
        "x": 0,
        "y": 0,
      },
      "a2": {
        "id": "a2",
        "index": "a2",
        "x": 100,
        "y": 700,
      },
    },
    "scale": 1,
    "size": "m",
    "spline": "line",
  },
  "rotation": 0,
  "type": "line",
  "typeName": "shape",
  "x": 150,
  "y": 150,
}
`;
